.ci/ci-fast-return.sh

```diff
@@ -0,0 +1,425 @@
+	results+=("check1 .*check2$")
+
+	# Check a pattern with no list does not fault.
+	repos+=("common")
+	patterns+=("empty_patterns")
+	results+=("")
+
+	# Check we do not fault looking for a pattern that is not defined.
+	repos+=("common")
+	patterns+=("undefined_patterns")
+	results+=("")
+
+	# Check we can read a named repo.
+	# Check single line expression entries work.
+	repos+=("documentation")
+	patterns+=("doc_single_entry")
+	results+=("single_pattern")
+
+	# Check we can handle multiple complex regexps.
+	repos+=("documentation")
+	patterns+=("doc_complex_patterns")
+	results+=("^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$ ^#?([a-f0-9]{6}|[a-f0-9]{3})$")
+
+	# Check we do not fail if trying to assess a repo that is not defined.
+	repos+=("nonrepo")
+	patterns+=("nonrepo_pattern")
+	results+=("")
+
+	count=0
+	for x in ${repos[@]}; do
+		echo " $count: Testing $x:${patterns[$count]} == '${results[$count]}'"
+
+		res=$(read_yaml "$unit_yamlfile" "${x}.${patterns[$count]}")
+		assertEquals "${results[$count]}" "$res"
+
+		count=$(( count+1 ))
+	done
+}
+
+# Unit test: Check the pattern matcher functions work as expected.
+testPatternMatcher() {
+	filenames="fred
+john
+mark
+doc.md"
+
+	matches=$(check_matches '.*\.md')
+	assertEquals "doc.md" "$matches"
+
+	# Check it also works with `-e` multiple expressions
+	matches=$(check_matches '-e notaname -e .*\.md')
+	assertEquals "doc.md" "$matches"
+
+	matches=$(check_matches 'mark')
+	assertEquals "mark" "$matches"
+
+	matches=$(check_matches '^.*$')
+	assertEquals "$filenames" "$matches"
+
+	matches=$(check_not_matches '.*\.md')
+	assertEquals "fred
+john
+mark" "$matches"
+
+	# check with multi patterns
+	matches=$(check_not_matches '-e .*\.md -e john')
+	assertEquals "fred
+mark" "$matches"
+
+
+	matches=$(check_not_matches 'mark')
+	assertEquals "fred
+john
+doc.md" "$matches"
+
+	matches=$(check_not_matches '^.*$')
+	assertEquals "" "$matches"
+
+}
+
+# Check if any of our own files have changed in this PR, and if so, run our own
+# unit tests...
+check_for_self_test() {
+	local ourfiles=""
+
+	ourfiles+="-e ^${script_gitname}$ "
+	ourfiles+="-e ^${yqfile_gitname}$ "
+	ourfiles+="-e ^${unit_yamlfile_gitname}$ "
+
+	res=$(check_matches $ourfiles)
+
+	if [ -n "$res" ]; then
+		local_info "file(s) [$res] modified - self running unit tests"
+		# Need to push the unit test output via the stdout mapped file descriptor
+		# so we don't corrupt our actual test function return values.
+		${cidir}/${script_name} test >&5
+	fi
+}
+
+# Run our self tests. Tests are written using the
+# github.com/kward/shunit2 library, and are encoded into functions starting
+# with the string 'test'.
+self_test() {
+	local_info "Running self tests"
+
+	local_info "Go get unit test framework"
+	go get -d github.com/kward/shunit2 || true
+	local_info "Run the unit tests"
+	# Sourcing the `shunit2` file automatically runs the unit tests in this file.
+	. ${GOPATH}/src/github.com/kward/shunit2/shunit2
```

Comment:
You could use a variable to avoid the repetition of `github.com/kward/shunit2`. A couple of other thoughts:

- I guess we always have go available but you could `curl` directly.
- This could be installed as a package instead by updating `.ci/setup_env_*.sh`.

Reply:
variable added.
The test will run so rarely, probably not worth adding it as a top level dep.